### PR TITLE
[ENH] FieldTrip Plugin for MNE-Scan

### LINKED ISSUE #399: Add Fieldtrip buffer support to MNE Scan

### PR DESCRIPTION
**Pull request by gabrielbmotta**
FtBuffer plugin interfaces with a ftbuffer and receives data from a neuromag machine. ghpages page documenting best use practices.

closes #399

### CODE REVIEW

File: applications/mne_scan/libs/scShared/Management/pluginmanager.cpp [hunk 1]
```diff
@@ -49,6 +49,7 @@
 #include <../plugins/neuromag/neuromag.h>
 #include <../plugins/babymeg/babymeg.h>
 #include <../plugins/natus/natus.h>
+//#include <../plugins/ftbuffer/ftbuffer.h>
```
Comment (LorenzE) [inline applications/mne_scan/libs/scShared/Management/pluginmanager.cpp]: the ft plugin can be commented in. only plugins which need closed source SDKs are commented out here.

File: applications/mne_scan/libs/scShared/scShared.pro [hunk 2]
```diff
@@ -68,6 +68,7 @@ contains(MNECPP_CONFIG, static) {
     QTPLUGIN += noisereduction
     QTPLUGIN += ssvepbci
     QTPLUGIN += neuronalconnectivity
+#    QTPLUGIN += ftbuffer
```
Comment (LorenzE) [inline applications/mne_scan/libs/scShared/scShared.pro]: same as above

File: applications/mne_scan/mne_scan/mne_scan.pro [hunk 3]
```diff
@@ -75,6 +75,7 @@ contains(MNECPP_CONFIG, static) {
     QTPLUGIN += ssvepbci
     QTPLUGIN += neuronalconnectivity
     QTPLUGIN += reference
+#    QTPLUGIN += ftbuffer
```
Comment (LorenzE) [inline applications/mne_scan/mne_scan/mne_scan.pro]: same as above

File: applications/mne_scan/plugins/ftbuffer/ftconnector.h [hunk 4]
```diff
@@ -0,0 +1,373 @@
...
+     */
+    int totalBuffSamples();
+
+    int                                     m_iNumSamples;                          /**< Number of samples we've read from the buffer */
+    int                                     m_iNumNewSamples;                       /**< Number of total samples (read and unread) in the buffer */
+    int                                     m_iMsgSamples;                          /**< Number of samples in the latest buffer transmission receied */
+    int                                     m_iNumChannels;                         /**< Number of channels in the buffer data */
+    int                                     m_iDataType;                            /**< Type of data in the buffer */
+    int                                     m_iNeuromagHeader;                      /**< Size of neuromag header chunk */
+    int                                     m_iPort             = 1972;             /**< Port where the ft bufferis found */
```
Comment (LorenzE) [inline applications/mne_scan/plugins/ftbuffer/ftconnector.h]: Please perform init in constructor only, see [here](https://stackoverflow.com/questions/28413154/c-initializing-variables-in-header-vs-with-constructor/28414747). We decided on using the "init in constructor" method.

File: applications/mne_scan/plugins/plugins.pro [hunk 5]
```diff
@@ -43,6 +43,7 @@ TEMPLATE = subdirs
 #Sensors
 SUBDIRS += \
     fiffsimulator \
+    ftbuffer \ #needs fieldtrip buffer source code
```
Comment (LorenzE) [inline applications/mne_scan/plugins/plugins.pro]: I think the comment `#needs fieldtrip buffer source code` can be removed, right?

File: doc/gh-pages/pages/development/ftbufferplugin.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
+---
+title: MNE-SCAN FieldTrip Buffer Plugin
+parent: Develop
+nav_order: 7
+---
+
+# MNE-SCAN FieldTrip Buffer Plugin
```
Comment (LorenzE) [inline doc/gh-pages/pages/development/ftbufferplugin.md]: Please change title to `# MNE-SCAN FieldTrip Buffer`

File: examples/ex_ft_buffer/ex_ft_buffer.pro [hunk 7]
```diff
@@ -0,0 +1,138 @@
+#--------------------------------------------------------------------------------------------------------------
```
Comment (LorenzE) [inline examples/ex_ft_buffer/ex_ft_buffer.pro]: Can you briefly elaborate if it makes sense to keep the example? Does it use the same code as the ft plugin in MNE Scan?

Comment (gabrielbmotta) [reply examples/ex_ft_buffer/ex_ft_buffer.pro]: It does not. I meant to delete it but forgot to. Newest push will remove it.

File: applications/mne_scan/plugins/ftbuffer/FormFiles/ftbuffersetupwidget.cpp [hunk 8]
```diff
@@ -0,0 +1,110 @@
+//=============================================================================================================
```
Comment (LorenzE) [inline applications/mne_scan/plugins/ftbuffer/FormFiles/ftbuffersetupwidget.cpp]: I just compiled your branch on win and mac and came across the `FtBufferSetupWidget`. Do we need this widget? If not please remove the widget and its action. The same applies for `FtBufferAboutWidget`.

File: applications/mne_scan/plugins/ftbuffer/FormFiles/ftbufferabout.ui [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+   </property>
+   <property name="html">
+    <string>&lt;!DOCTYPE HTML PUBLIC &quot;-//W3C//DTD HTML 4.0//EN&quot; &quot;http://www.w3.org/TR/REC-html40/strict.dtd&quot;&gt;
+&lt;html&gt;&lt;head&gt;&lt;meta name=&quot;qrichtext&quot; content=&quot;1&quot; /&gt;&lt;style type=&quot;text/css&quot;&gt;
+p, li { white-space: pre-wrap; }
+&lt;/style&gt;&lt;/head&gt;&lt;body style=&quot; font-family:'Sans'; font-size:10pt; font-weight:400; font-style:normal;&quot;&gt;
+&lt;table border=&quot;0&quot; style=&quot;-qt-table-type: root; margin-top:4px; margin-bottom:4px; margin-left:4px; margin-right:4px;&quot;&gt;
+&lt;tr&gt;
+&lt;td style=&quot;border: none;&quot;&gt;
+&lt;p style=&quot; margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;&quot;&gt;&lt;span style=&quot; font-family:'Sans';&quot;&gt;Copyright © 2020 Matti Hamalainen, Lorenz Esch. All rights reserved.&lt;/span&gt;&lt;/p&gt;&lt;/td&gt;&lt;/tr&gt;&lt;/table&gt;&lt;/body&gt;&lt;/html&gt;</string>
```
Comment (LorenzE) [inline applications/mne_scan/plugins/ftbuffer/FormFiles/ftbufferabout.ui]: This should be `Gabriel B Motta and Lorenz Esch`

**Comment by LorenzE** [comment]
Test with VV on Mac, Linux and Win10 were successful. Some notes:
- The ftbuffer plugin crashes when starting without parameters.
- Disconnecting from neuromag2ft in MNE Scan and leaving the server run for a while breaks MNE Scan when trying to reconnect. Problem might be that we are trying to receive all the data accumulated since the last disconnect. Maybe only get the latest data block instead when trying to reconnect.
- Reduce amount of console output for received data block in MNE Scan.
- Does neuromag2ft has to be started before the measurement is started on the acquisition computer? This might be good information to add to the docu.

**Pull request merged**